### PR TITLE
TIP-1173: Implement a ReadValueCollection and remove the other Collection Interfaces

### PR DESCRIPTION
**Pull request by anaelChardan**
**Description (for Contributor and Core Developer)**

This PR split the reading side and the writing side of the ValueCollectionInterface
It removes the ValueCollectionInterface which is useless (I've added a sed to be able to change that for all customers).

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | OK
| Added legacy Behats               | -
| Added acceptance tests            | -
| Added integration tests           | -
| Changelog updated                 | OK
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/Pim/Enrichment/Component/Product/Grid/ReadModel/Row.php [hunk 1]
```diff
@@ -5,7 +5,7 @@
 namespace Akeneo\Pim\Enrichment\Component\Product\Grid\ReadModel;
 
 use Akeneo\Pim\Enrichment\Component\Product\Model\ValueCollection;
-use Akeneo\Pim\Enrichment\Component\Product\Model\ValueCollectionInterface;
```
Comment (anaelChardan) [inline src/Akeneo/Pim/Enrichment/Component/Product/Grid/ReadModel/Row.php]: TODO: We'll see that afterward

File: src/Akeneo/Pim/Enrichment/Component/Product/Model/IndexedValueCollectionInterface.php [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+ * It does not extends \ArrayAccess as indexation could be performed internally
+ * for performance reasons.
+ *
+ * Highly inspired by \Doctrine\Common\Collection.
+ *
+ * @author    Julien Janvier <j.janvier@gmail.com>
+ * @copyright 2017 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+interface IndexedValueCollectionInterface extends ValueCollectionInterface
```
Comment (anaelChardan) [inline src/Akeneo/Pim/Enrichment/Component/Product/Model/IndexedValueCollectionInterface.php]: Could be named WriteValueCollectionInterface

File: src/Akeneo/Pim/Enrichment/Bundle/Filter/ProductValueChannelFilter.php [hunk 3]
```diff
@@ -2,8 +2,8 @@
 
 namespace Akeneo\Pim\Enrichment\Bundle\Filter;
 
-use Akeneo\Pim\Enrichment\Component\Product\Model\ValueCollectionInterface;
 use Akeneo\Pim\Enrichment\Component\Product\Model\ValueInterface;
+use Akeneo\Pim\Enrichment\Component\Product\Model\WriteValueCollectionInterface as ValueCollectionInterface;
```
Comment (pierallard) [inline src/Akeneo/Pim/Enrichment/Bundle/Filter/ProductValueChannelFilter.php]: I don't get why you do an alias in each class.

File: src/Akeneo/Pim/Enrichment/Component/Product/Model/ProductModel.php [hunk 4]
```diff
@@ -643,13 +643,13 @@ public function __toString()
 
     /**
      * @param EntityWithFamilyVariantInterface $entity
-     * @param ValueCollectionInterface         $valueCollection
+     * @param WriteValueCollectionInterface  $valueCollection
```
Comment (pierallard) [inline src/Akeneo/Pim/Enrichment/Component/Product/Model/ProductModel.php]: Some men just want to watch the world burn

File: src/Akeneo/Pim/Enrichment/Component/Product/Model/ReadValueCollection.php [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+ * @author    Anael Chardan <anael.chardan@akeneo.com>
+ * @copyright 2019 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+class ReadValueCollection implements ValueCollectionInterface
+{
+    /** @var ValueInterface[] */
+    private $values;
+
+    private $attributeCodes = [];
```
Comment (pierallard) [inline src/Akeneo/Pim/Enrichment/Component/Product/Model/ReadValueCollection.php]: typehint

File: src/Akeneo/Pim/Enrichment/Bundle/Filter/ProductValueChannelFilter.php (hunk 3 above)
Comment (anaelChardan) [reply src/Akeneo/Pim/Enrichment/Bundle/Filter/ProductValueChannelFilter.php]: To avoid to rename everywhere

**Pull request merged**